### ISSUE TITLE
为什么需要对logits进行右移

### ISSUE THREAD

**Issue by xinpengzz** [HKUNLP/Dream#31]
您好，请问在这里为什么需要对logits进行右移呢？
```python
for i in range(steps):
    mask_index = (x == mask_token_id)
    logits = self(x, attention_mask, tok_idx).logits
    logits = torch.cat([logits[:,:1], logits[:, :-1]], dim=1) # why?
```
https://huggingface.co/Dream-org/Dream-v0-Base-7B/blob/main/generation_utils.py#L409-L412

**Comment by dhruvdcoder** [comment]
I have the same question.

**Comment by jiacheng-ye** [comment]
Hi there, great question. As in Figure (b) in the training part of https://hkunlp.github.io/blog/2025/dream/, we use shift prediction, so here we re-shift the prediction to align with labels for the loss calculation. One can also shift the labels instead of logits.

**Comment by stas-sl** [comment]
That's quite an interesting detail - I hadn't come across it before. I used to think that masked language models (MLM), like the earlier BERT or the newer LLaDA, were trained solely to predict masked tokens, rather than the next token as in AR models. But in your model, you're predicting the next token after a masked one, which seems like a hybrid of both approaches. It feels a bit odd at first, but as you explained, it was likely done to enable weight reuse from AR models and, I suppose, to also allow the model to function as an AR model during decoding. So yes, it's definitely interesting - just a bit unconventional.
